rust/agama-software/src/callbacks/commit_download.rs

```diff
@@ -0,0 +1,78 @@
+use agama_utils::{actor::Handler, api::question::QuestionSpec, progress, question};
+use gettextrs::gettext;
+use tokio::runtime::Handle;
+use zypp_agama::callbacks::pkg_download::{Callback, DownloadError};
+
+#[derive(Clone)]
+pub struct CommitDownload {
+    progress: Handler<progress::Service>,
+    questions: Handler<question::Service>,
+}
+
+impl CommitDownload {
+    pub fn new(
+        progress: Handler<progress::Service>,
+        questions: Handler<question::Service>,
+    ) -> Self {
+        Self {
+            progress,
+            questions,
+        }
+    }
+}
+
+impl Callback for CommitDownload {
+    fn start_preload(&self) {
+        // TODO: report progress that we start preloading packages
+        tracing::info!("Start preload");
+    }
+
+    fn problem(
+        &self,
+        name: &str,
+        error: DownloadError,
+        description: &str,
+    ) -> zypp_agama::callbacks::ProblemResponse {
+        // TODO: make it generic for any problemResponse questions
+        let labels = [gettext("Retry"), gettext("Ignore")];
+        let actions = [
+            ("Retry", labels[0].as_str()),
+            ("Ignore", labels[1].as_str()),
+        ];
+        let error_str = error.to_string();
+        let data = [("package", name), ("error_code", error_str.as_str())];
+        let question = QuestionSpec::new(description, "software.package_error.provide_error")
+            .with_actions(&actions)
+            .with_data(&data);
+        let result = Handle::current().block_on(async move {
```

Comment:
NP: I feel the bottom half of the fn is boilerplate and we will want to factor it out as more callbacks appear